src/main/kotlin/me/zeroeightsix/kami/feature/module/Notebot.kt

```diff
@@ -0,0 +1,151 @@
+    @Setting
+    var mode = NotebotMode.DISCOVER
+
+    @Setting(comment = "Name of .mid file in the songs folder of your .minecraft")
+    var songName = "tetris_theme.mid"
+
+    @Setting
+    var hatAlwaysAsFSharp = true
+
+    @Setting(name = "Channel 0")
+    var ChannelZero = Instrument.HAT
+
+    @Setting(name = "Channel 1")
+    var ChannelOne = Instrument.HAT
+
+    @Setting(name = "Channel 2")
+    var ChannelTwo = Instrument.HAT
+
+    @Setting(name = "Channel 3")
+    var ChannelThree = Instrument.HAT
+
+    @Setting(name = "Channel 4")
+    var ChannelFour = Instrument.HAT
+
+    private var channelsArray = arrayOf(ChannelZero, ChannelOne, ChannelTwo, ChannelThree, ChannelFour)
+
+    private var lastNote: Long = 0L
+    private var elapsed: Long = 0L
+
+    private var noteSequence: TreeMap<Long, ArrayList<Note>> = TreeMap()
+    var map: InstrumentMap = InstrumentMap()
+
+    enum class NotebotMode {
+        DISCOVER,
+        LOAD,
+        PLAY
+    }
+
+    @EventHandler
+    val listener = Listener<TickEvent.InGame>({
+        val player = it.player
+        val world = it.world
+
+        when (mode) {
+            NotebotMode.DISCOVER -> {
+                for (x in -4..4) for (y in -4..4) for (z in -4..4) {
+                    val pos = player.pos?.plus(Vec3d(x.toDouble(), y.toDouble(), z.toDouble())) ?: break
+                    if (world.getBlockState(BlockPos(pos))?.block == Blocks.NOTE_BLOCK) {
+                        val noteBlock = world.getBlockState(BlockPos(pos))
+                        snackbarMessage(
+                            player, "" + noteBlock.get(INSTRUMENT) + "[" + noteBlock.get(NOTE) + "]"
+                        )
+                        map.add(noteBlock.get(INSTRUMENT), noteBlock.get(NOTE), BlockPos(pos))
+                    }
+                }
+            }
+            NotebotMode.LOAD -> {
+                noteSequence = MidiParser.parse("songs" + File.separator + songName, 5)
+                if (noteSequence.isEmpty())
+                    snackbarMessage(player, "Song not found: " + "songs" + File.separator + songName)
+                else
+                    snackbarMessage(player, "Song Found: " + "songs" + File.separator + songName)
+                lastNote = System.currentTimeMillis()
+                elapsed = 0
+            }
+            NotebotMode.PLAY -> {
+                if (!player.isCreative) {
+                    while (noteSequence.isNotEmpty() && noteSequence.firstKey() <= elapsed) {
+                        playNotes(noteSequence.pollFirstEntry().value, player, world)
+                    }
+                    elapsed += System.currentTimeMillis() - lastNote
+                    lastNote = System.currentTimeMillis()
+                } else snackbarMessage(player, "You are in creative mode and cannot play music.")
+            }
+        }
+    })
+
+    private fun playNotes(notes: ArrayList<Note>, player: ClientPlayerEntity, world: ClientWorld) {
```

Comment:
```suggestion
    private fun playNotes(notes: List<Note>, player: ClientPlayerEntity, world: World) {
```
To make this work, also change the signature of `playBlock` and `getIrreplaceableNeighbour`.